### PR TITLE
Request classification proto

### PR DESCRIPTION
**Pull request by mandarjog**
Add: proto definition for request classification.
Add: two examples for use.

Note: the configuration must be json at the point of delivery, not yaml.
Implementation not done yet.

### CODE REVIEW

**Comment by mandarjog** [comment]
@douglas-reid @kyessenov @nrjpoddar PTAL the proto definitions and examples.

**Comment by mandarjog** [comment]
@bianpengyuan look like a new lint rule?

‘ test/envoye2e/driver/fake_stackdriver.go:48:2: ST1003: struct field listTsResp should be listTSResp (stylecheck)
74
	listTsResp   monitoringpb.ListTimeSeriesResponse’

File: extensions/classify/config.proto [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+// Dimension configuration produces the named dimension with a value based
+// on the match clause in the value. 
+// The value matches are processed sequentially and the first value to match is assigned.
+message Dimension {
+  // For example `operation` is a dimension.
+  string name = 1;
+
+  // At most one of the following values is assigned to the dimension. 
+  repeated Value values = 2;
```
Comment (douglas-reid) [inline extensions/classify/config.proto:45]: I wonder: would this be clearer if it was `repeated Classifier classifiers = 2` ?

File: extensions/classify/config.proto [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+// Match condition based on request and response information.
+// Every match condition must match for the result to be true.
+message Match {
+  // invert the match condition.
+  bool invert = 1;
+
+  // Match uri.
+  StringMatch uri = 2;
```
Comment (douglas-reid) [inline extensions/classify/config.proto:66]: Should this be split into two: `host` and `path`  ?  Or even three: `host`, `path`, `query_params` ?

File: extensions/classify/config.proto [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  // $hide_from_docs
+  map<string, StringMatch> query_params = 7;
+
+  // Match response code.
+  IntMatch response_code = 8;
+
+  // Match request size.
+  // Not implemented.
+  // $hide_from_docs
+  IntMatch request_size = 9;
```
Comment (douglas-reid) [inline extensions/classify/config.proto:95]: if we are classifying by size, will we ever want to enable classifying by durations ?

File: extensions/classify/config.proto [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+message Match {
+  // invert the match condition.
+  bool invert = 1;
+
+  // Match uri.
+  StringMatch uri = 2;
+
+  // Match scheme: http or https.
+  string scheme = 3;
+
```
Comment (douglas-reid) [inline extensions/classify/config.proto:70]: do we care about classifying by HTTP version at all?

File: extensions/classify/config.proto [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+// Value assigned to the dimension if the request matches.
+message Value {
+  // Assign `value` to the enclosing dimension if match is true.
+  string value = 1;
+
+  // Match specification.
+  // Empty match always matches, which can be used to specify a default value when placed
+  // at the end of the list.
+  Match match = 2;
```
Comment (douglas-reid) [inline extensions/classify/config.proto:56]: Do we have a way / plan to handle overlapping matches ?  Is it first to match is used?

File: extensions/classify/testdata/operation.json [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+              "regex": "^/reviews/\d*"
+            },
+            "method": "GET"
+          }
+        },
+        {
+          "value": "ListReviews",
+          "match": {
+            "uri": {
+              "exact": "/reviews/"
```
Comment (douglas-reid) [inline extensions/classify/testdata/operation.json:20]: in testing this: we should add a testcase for something like `/books/{id}/chapter/{id}`

File: extensions/classify/config.proto [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  // Less than or equal to this number.
+  int64 le = 2;
+}
+
+// PluginPhase determines when the classification is performed.
+enum PluginPhase {
+  // Classification is performed after the response is sent back to the client.
+  // All information about the request is available in this phase,
+  // and classification does not add to the latency observed by the client.
+  LOG = 0;
```
Comment (douglas-reid) [inline extensions/classify/config.proto:147]: `LOG` and `REQUEST` seem very envoy-specific.  Should we use something like:

`AFTER_RESPONSE`, `ON_REQUEST`, `ON_RESPONSE`, or something?

Comment (douglas-reid) [inline extensions/classify/config.proto:147]: Do we care about style guide recommendations here?

```
PHASE_UNSPECIFIED = 0; // Defaults to after response
ON_REQUEST_PHASE = 1;
AFTER_RESPONSE_PHASE = 2;
```

File: extensions/classify/config.proto [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+message IntRange {
+  // Greater than or equal to this number.
+  int64 ge = 1;
+
+  // Less than or equal to this number.
+  int64 le = 2;
+}
+
+// PluginPhase determines when the classification is performed.
+enum PluginPhase {
```
Comment (douglas-reid) [inline extensions/classify/config.proto:143]: If we don't call this `ClassificationPhase`, I'd be in favor of dropping `Plugin`, and having this just be `Phase`.

File: extensions/classify/config.proto [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+    // exact Int match
+    int64 exact = 1;
+
+    // Range match
+    IntRange range = 2;
+  }
+}
+
+// IntRange matches a range of integers.
+message IntRange {
```
Comment (douglas-reid) [inline extensions/classify/config.proto:134]: What do you think about using `from` and `to` or `minimum` and `maximum` instead of `ge` and `le` here ?

Comment (douglas-reid) [inline extensions/classify/config.proto:134]: As this is the only range we support, should this just be `Range` ?

File: extensions/classify/config.proto [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+syntax = "proto3";
+
+package classify;
```
Comment (kyessenov) [inline extensions/classify/config.proto:18]: Please use `istio` in the package prefix, since this is probably istio-specific for now.
Meta-comment: no need to use proto for extension configuration. You could just use JSON/OpenAPI.

File: extensions/classify/config.proto [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+
+// Configuration to classify http traffic.
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply extensions/classify/config.proto:21]: nit: capital `HTTP`.

File: extensions/classify/config.proto [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+// Traffic classification can be done on mutiple dimensions using request and
+// response headers and other measurements. 
+// The plugin must be configured as request, response or log based on
```
Comment (kyessenov) [inline extensions/classify/config.proto:24]: There is no response based plugins as far as I know. All plugins implement the same callbacks, except for access log based ones.

File: extensions/classify/config.proto [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+// which other plugins need to consume the classfication results.
+// For exmaple: If Stats plugin uses `operation` dimension produced by the classification
+// plugin, then it must be executed in LOG phase *before* the Stats plugin.
+message PluginConfig {
+  // next id: 3
+  // Plugin phase when classification is done.
+  PluginPhase phase = 1;
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply extensions/classify/config.proto:31]: Is it really expensive to run this in onRequestHeaders always? It might be pre-mature optimization to try to move it to the logging phase.

File: extensions/classify/config.proto (hunk 5 above)
Comment (kyessenov) [reply extensions/classify/config.proto:56]: This feels like yet-another AST for a condition? Why not just use CEL expression instead of creating one?

File: extensions/classify/config.proto (hunk 8 above)
Comment (kyessenov) [reply extensions/classify/config.proto:143]: I wonder if this could be implemented in a way that is transparent to the phase. Maybe some kind of lazily computed property.

File: extensions/classify/config.proto (hunk 7 above)
Comment (mandarjog) [reply extensions/classify/config.proto:147]: good point, updated names.

File: extensions/classify/config.proto (hunk 13 above)
[the remark this answers is not in the document]
Comment (mandarjog) [reply extensions/classify/config.proto]: It is not an optimization.

1. It depends on what information is available when. (response codes, durations are not available on_request)
2. Which downstream plugin needs this information.

For example, you can have a `operation` classifier deployed as `ON_REQUEST` which can be used by authz and subsequently by stats.
If you also want to use responseClassification it should happen `before` stats filter gets called.

@kyessenov are you suggesting that we do it automagically, and the implementation decides when classification takes place?

File: extensions/classify/config.proto [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+
+  // Match specification.
+  // Empty match always matches, which can be used to specify a default value when placed
+  // at the end of the list.
+  Match match = 2;
+}
+
+// Match condition based on request and response information.
+// Every match condition must match for the result to be true.
+message Match {
```
Comment (mandarjog) [inline extensions/classify/config.proto:62]: @kyessenov are you suggesting that we replace Match message with a CEL expression?
I would still have a list of those which would be more readable.

I think all the headers are already available in the CEL, are any measurements like response duration available there?
We can always add.

File: extensions/classify/config.proto (hunk 12 above)
Comment (nrjpoddar) [reply extensions/classify/config.proto:25]: Lines 25:26 is quite hard to follow, may be add this:

This filter generates attributes based on the configured classification and request/response being processed. These attributes can be consumed by downstream filters in the chain. For e.g. Stats plugin uses `operation` dimension produced by the classification plugin, then it must be executed in LOG phase *before* the Stats plugin.

File: extensions/classify/config.proto [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+  Phase phase = 1;
+
+  // Request classification along multiple dimensions.
+  repeated Dimension dimensions = 2;
+}
+
+// Dimension configuration produces the named dimension with a value based
+// on the match clause in the value. 
+// The value matches are processed sequentially and the first value to match is assigned.
+message Dimension {
```
Comment (nrjpoddar) [inline extensions/classify/config.proto:41]: Dimension encapsulates the configuration of a named attribute which is emitted by the filter if the conditions are met.

File: extensions/classify/config.proto [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+  // For example `operation` is a dimension.
+  string name = 1;
+
+  // At most one of the following values is assigned to the dimension. 
+  repeated Value values = 2;
```
Comment (nrjpoddar) [inline extensions/classify/config.proto:46]: The values are evaluated in order and the first condition to match wins.

File: extensions/classify/config.proto [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+// Value assigned to the dimension if the request matches.
+message Value {
+  // Assign `value` to the enclosing dimension if match is true.
```
Comment (nrjpoddar) [inline extensions/classify/config.proto:51]: Value of the dimension if the match condition is met

File: extensions/classify/config.proto [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+  // Request classification along multiple dimensions.
+  repeated Dimension dimensions = 2;
+}
+
+// Dimension configuration produces the named dimension with a value based
+// on the match clause in the value. 
+// The value matches are processed sequentially and the first value to match is assigned.
+message Dimension {
+  // For example `operation` is a dimension.
+  string name = 1;
```
Comment (nrjpoddar) [inline extensions/classify/config.proto:43]: Name of the dimension for e.g. "operation"

File: extensions/classify/config.proto [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+  // Match scheme: http, https or unset.
+  Scheme scheme = 3;
+
+  // Match method: GET, POST, HEAD, PUT, DELETE, ...
+  repeated string method = 4;
+
+  // Match request headers.
+  // Not implemented.
+  // $hide_from_docs
+  map<string, StringMatch> request_headers = 5;
```
Comment (nrjpoddar) [inline extensions/classify/config.proto:78]: Since you have to copy these fields from VirtualService API instead of importing, wdyt about using CEL and listing available attributes?

Have a copy of the API field is kind of a maintenance nightmare with accumulating tech debt of keeping parity.

File: extensions/classify/config.proto [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+enum Scheme {
+  // Matches any
+  UNSPECIFIED = 0;
+  HTTP = 1;
+  HTTPS = 2;
+}
+
+// Phase determines when the classification is performed.
```
Comment (nrjpoddar) [inline extensions/classify/config.proto:180]: Does Phase capture intent of the user or actually places the filter at the right place i.e. do users still need to create EnvoyFilter with the current placement settings for this new filter?

File: extensions/classify/testdata/operation.json [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+            },
+            "method": "GET"
+          }
+        },
+        {
+          "value": "CreateReview",
+          "match": {
+            "uri": {
+              "exact": "/reviews/"
```
Comment (gargnupur) [inline extensions/classify/testdata/operation.json:29]: nit: add a match test for prefix?

File: extensions/classify/config.proto (hunk 13 above)
[the remark this answers is not in the document]
Comment (kyessenov) [reply extensions/classify/config.proto]: You should add presence indicator for information that is missing. You don't always have those attributes even on logging callback (e.g. abrupt request termination).

File: extensions/classify/config.proto (hunk 2 above)
Comment (kyessenov) [reply extensions/classify/config.proto]: We need to be very precise what this means if this is going to be used for authz. Path, canonicalized path, query params all make a huge difference and can cause a CVE if there's some misunderstanding.

File: extensions/classify/config.proto (hunk 14 above)
Comment (kyessenov) [reply extensions/classify/config.proto:62]: Yeah, you can follow the same pattern as authz in Envoy. Have an AST predicate with a CEL expression for more complicated logic (ORs, ANDs, functions).

File: extensions/classify/config.proto (hunk 7 above)
Comment (kyessenov) [reply extensions/classify/config.proto]: h2 allows bi-directional streaming. So there's no clear boundary between when request ends and response starts.

File: extensions/classify/config.proto (hunk 19 above)
Comment (mandarjog) [reply extensions/classify/config.proto:78]: I will try out a CEL only api and see how it works. The proto defn will be more. Concise and kuat and I have been discussing that.
The user experience and validation is the issue.

I am not worried about maintaining parity with virtual service.
Neither APIs matching properties are expected to change often.

File: extensions/classify/config.proto (hunk 20 above)
Comment (mandarjog) [reply extensions/classify/config.proto:180]: User needs to place the filter correctly in relation to other filters.
Phase dictates the filter callback when classification is performed.

File: extensions/classify/config.proto (hunk 19 above)
Comment (mandarjog) [reply extensions/classify/config.proto:78]: @kyessenov specifying cel in json with escaping “ will be a pain so we need envoy filter api to do the conversations.

File: extensions/classify/config.proto (hunk 20 above)
Comment (nrjpoddar) [reply extensions/classify/config.proto:180]: Isn't that too low level for any user?

Why can't we have the Phase fixed for the most common use cases as most of the end users won't understand this.

**Comment by nrjpoddar** [comment]
Just as a thought exercise, how would we make this protocol agnostic. This ties a bit into my RFC for meta routing. Since HTTP mostly is a superset of attributes and your API is already abstract, the only thing making it more specific to HTTP is using the name "Request/Response"

@douglas-reid @mandarjog FYI

**Comment by mandarjog** [comment]
Peer attributes, self attributes and protocol attributes are the 3 types of attributes available. I am making an alternate PR with CEL only in that case all attributes are avl inside CEL, but there will be no request response nomenclature in the proto itself.

For kafka: we could have new cel attributes that you can use.

File: extensions/classify/config.proto (hunk 20 above)
Comment (mandarjog) [reply extensions/classify/config.proto:180]: You cannot have a fixed phase. If I want to use the output of classification for responseClass in the stats filter then the classification must be performed in LOG phase when responseCodes are available.

Alternatively if you want to use the classification in an authz filter you cannot perform classification in the LOG phase, it will be too late and authz will not have the request classification that is needed.

We can do some magic to figure this out, I think we should start simple and explicit .

Comment (kyessenov) [reply extensions/classify/config.proto:180]: Can the filter guess based on its position? It's not a good experience to both specify the filter position in EnvoyFilter, and then specify it again in the phase of the filter config.

**Comment by mandarjog** [comment]
@kyessenov @douglas-reid @nrjpoddar @gargnupur I have created a new PR https://github.com/istio/proxy/pull/2801 which is CEL based, please review.

**Comment by mandarjog** [comment]
closing this in favour of #2801
